### PR TITLE
defensive programming around missing dependency files during an update

### LINKED ISSUE #6487: Review `required_files_in?` necessity
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Code improvement description

We have a `required_files_in?` as part of our public fetcher API: https://github.com/dependabot/dependabot-core/blob/5024751db27968f00c769ed93e34243e268b780f/common/lib/dependabot/file_fetchers/README.md#public-api

But only the Bundler ecosystem seems to be actually using it? Ecosystems seem to normally override `fetch_files` (a private method called by the public `files` method) and raise an explicit error from there when a required file is not found.

Keeping these while being unused causes additional burden to both contributors and maintainers. See https://github.com/dependabot/dependabot-core/pull/6248#discussion_r1081860748.

We should evaluate whether we should deprecate this method, or start actually using it.

### PR DESCRIPTION
**Pull request by jakecoffman**
We trust that the ecosystems are returning an array of dependencies but we never really check if they are and we continue with the update. It's possible due to an error that the array is empty, so let's check for that.

Also during a grouped update we filter out based on the directory, which should never result in an empty array but just in case it does let's check for that. 

Edit: Also noticed a while back that the Updater never calls `reqiured_files_in?`. I think some ecosystems have setup explicit checks that the required files were fetched, but others assume the Updater code calls `required_files_in?`. So I went ahead and called it in the Base FileFetcher's `files` method so we can simplify that check across the board. fixes #6487

### CODE REVIEW

File: github_actions/lib/dependabot/github_actions/file_fetcher.rb [hunk 1]
```diff
@@ -11,10 +11,8 @@ class FileFetcher < Dependabot::FileFetchers::Base
       extend T::Sig
       extend T::Helpers
 
-      FILENAME_PATTERN = /^(\.github|action.ya?ml)$/
-
       def self.required_files_in?(filenames)
-        filenames.any? { |f| f.match?(FILENAME_PATTERN) }
```
Comment (abdulapopoola) [inline github_actions/lib/dependabot/github_actions/file_fetcher.rb:17]: any reason to remove the constant?

Comment (jakecoffman) [reply github_actions/lib/dependabot/github_actions/file_fetcher.rb:17]: No reason, I put it back.

File: common/lib/dependabot/file_fetchers/base.rb [hunk 2]
```diff
@@ -104,7 +108,16 @@ def target_branch
 
       sig { returns(T::Array[DependencyFile]) }
       def files
-        @files ||= T.let(fetch_files, T.nilable(T::Array[DependencyFile]))
+        return T.must(@files) if defined? @files
```
Comment (deivid-rodriguez) [inline common/lib/dependabot/file_fetchers/base.rb:111]: Are both this and the `T.must` in the last line of the method needed?

**Review by deivid-rodriguez — APPROVED** [review]
This is really great! It makes file fetchers easier and with a better default behavior!

Also this PR closes https://github.com/dependabot/dependabot-core/issues/6487 😃.

File: common/lib/dependabot/file_fetchers/base.rb (hunk 2 above)
Comment (jakecoffman) [reply common/lib/dependabot/file_fetchers/base.rb:111]: Yes but I think I can clean this up if I move the `@files` creation to the constructor.

Comment (jakecoffman) [reply common/lib/dependabot/file_fetchers/base.rb]: It doesn't seem to be the Ruby style (at least the Dependabot style?) to initialize all variables in the constructor but it makes using types easier.

**Review by brbayes-msft — APPROVED** [review]

**Comment by rvelusamy** [comment]
This PR is introducing an issue for https://github.com/dependabot/dependabot-core/issues/8867. Could someone check this?

**Pull request merged**